### ISSUE TITLE
Incorrect assumptions about SDK-v2 TypeMap encoding

### ISSUE THREAD

**Issue by t0yv0** [pulumi/pulumi-terraform-bridge#1828]
Randomized conformance testing has led me to an important corner case for encoding object types in SDKv2 that I
previously misunderstood, it is unfortunately possible that the bridge codebase does not do the right thing here and we
need to take a look and refine the assumptions.

The key code is here:  https://github.com/hashicorp/terraform-plugin-sdk/blob/main/helper/schema/core_schema_test.go#L220

Previously I assumed that to encode an object type or a single-nested TF block in schemav2 one would use
`Type=schema.TypeMap` and `Elem=*schema.Resource{}`.

However it appears that based on the evidence above, TF in this case would DISCARD the schema supplied in
`&schema.Resource` entirely and assume this type to be a `Map[string]`. The cited reason is for compatibility with TF 0.11...

It remains unclear to me how SDKv2 providers encode object types then. It seems from e.g. assumeRoleSchema() in AWS that they use `Type=schema.TypeList` and `Elem=*schema.Resource{}` and `MaxItems: 1`. So that MaxItems is required simply to encode object types. This may be the logical conclusion here.

In terms of the way forward:

- Revise the documentation on schemashim.Elem()
- Ensure that tfgen decides on Map[string] for this case
- Ensure tfgen errors if the user supplies a discarded Resource with TypeMap
  - if the upstream provider supplies a Resource, allow the bridge author to ignore the error
  - otherwise fail hard, we should not have this
- Scan for all runtime usage of the schema for this case
  - tests for makeTerraformInput
  - tests for makeTerraformOutput
  - fix PlanResourceChange code and type inference
  - possibly more

CC @iwahbe @VenelinMartinov

**Comment by VenelinMartinov** [comment]
`CastToTypeObject` in https://github.com/pulumi/pulumi-terraform-bridge/blob/89906ef933db60a27cc1dbe2bd2f6ce2278da8da/pkg/tfshim/util/types.go#L35 still suffers from this.

https://github.com/pulumi/pulumi-terraform-bridge/issues/2180 and https://github.com/pulumi/pulumi-terraform-bridge/issues/2185 look like consequensces.

Should we revisit this?

**Comment by t0yv0** [comment]
Yes, M109 perhaps. I did a light pass adjusting docu on Elem() but didn't finish auditing all places that use Elem() to make decisions.
